libclc/generic/libspirv/atomic/atomic_store.cl

```diff
@@ -29,26 +29,27 @@ _Z19__spirv_AtomicStorePU3AS3fN5__spv5Scope4FlagENS1_19MemorySemanticsMask4FlagE
-      __clc__atomic_##PREFIX##store_##AS##_##BYTE_SIZE##_seq_cst(p, val);                                                     \
-    } else {                                                                                                                  \
-      __clc__atomic_##PREFIX##store_##AS##_##BYTE_SIZE##_unordered(p, val);                                                   \
-    }                                                                                                                         \
+#define IMPL(TYPE, TYPE_MANGLED, AS, AS_MANGLED, SUB, PREFIX, BYTE_SIZE)                                                          \
+  FDECL(TYPE, PREFIX, AS, BYTE_SIZE, unordered)                                                                                   \
+  FDECL(TYPE, PREFIX, AS, BYTE_SIZE, release)                                                                                     \
+  FDECL(TYPE, PREFIX, AS, BYTE_SIZE, seq_cst)                                                                                     \
+  _CLC_DEF void                                                                                                                   \
+      _Z19__spirv_AtomicStoreP##AS_MANGLED##TYPE_MANGLED##N5__spv5Scope4FlagENS##SUB##_19MemorySemanticsMask4FlagE##TYPE_MANGLED( \
+          volatile AS TYPE *p, enum Scope scope,                                                                                  \
+          enum MemorySemanticsMask semantics, TYPE val) {                                                                         \
+    if (semantics == Release) {                                                                                                   \
+      __clc__atomic_##PREFIX##store_##AS##_##BYTE_SIZE##_release(p, val);                                                         \
+    } else if (semantics == SequentiallyConsistent) {                                                                             \
+      __clc__atomic_##PREFIX##store_##AS##_##BYTE_SIZE##_seq_cst(p, val);                                                         \
+    } else {                                                                                                                      \
+      __clc__atomic_##PREFIX##store_##AS##_##BYTE_SIZE##_unordered(p, val);                                                       \
+    }                                                                                                                             \
   }
 
-#define IMPL_AS(TYPE, TYPE_MANGLED, PREFIX, BYTE_SIZE) \
-IMPL(TYPE, TYPE_MANGLED, global, AS1, PREFIX, BYTE_SIZE) \
-IMPL(TYPE, TYPE_MANGLED, local, AS3, PREFIX, BYTE_SIZE)
+#define IMPL_AS(TYPE, TYPE_MANGLED, PREFIX, BYTE_SIZE)                         \
+  IMPL(TYPE, TYPE_MANGLED, global, U3AS1, 1, PREFIX, BYTE_SIZE)                \
+  IMPL(TYPE, TYPE_MANGLED, local, U3AS3, 1, PREFIX, BYTE_SIZE)                 \
+  IMPL(TYPE, TYPE_MANGLED, , , 0, PREFIX, BYTE_SIZE)
```

Comment:
Do we need to be careful about a generic target not supporting the generic address space? If the target doesn't support this address space, then would an undecorated pointer not technically be in the `private` address space? Then the mangling of the function name might not necessarily match the mangling of its pointer arguments.

Sorry if this also applies to the atomic load patch too - I may not have been thinking clearly :wink:

Reply:
Yeah I think you are right, currently I think that `Atomic*` operations in `generic` are used only by Native CPU, which also uses the `x86_64-unknown-linux` target folder. Do you think we should move all the `atomic/atomic_*.cl` files there and leave only the `loadstore_helpers` LLVM files in `generic` ?

Reply:
Good question. My instinct is to keep generic things in the generic folder, even if in practice only one target uses them. I suppose we could argue about whether hard-coding address spaces is truly generic, but I'm sure we can fix that somehow.

So unless we've got no way of expressing these builtins in a generic way, I vote to keep them here. We can use the `_CLC_GENERIC_AS_SUPPORTED` (not merged yet - see https://github.com/intel/llvm/pull/13423) define to keep it as generic as possible.

Reply:
Ok cool, I'll wait for https://github.com/intel/llvm/pull/13423 to get merged and use that macro for this PR (and the load too)

Reply:
Hi @frasercrmck, I've changed the PR to use `_CLC_GENERIC_AS_SUPPORTED`, can you take another look? Thanks